### ISSUE TITLE
Inconsistencies between Pyright and Pylance

### ISSUE THREAD

**Issue by albireox** [microsoft/pylance-release#1471]
I use Pylance with VSCode to check my code typing. On the repository side, I have a GitHub Action that updates Pyright to the latest version (using NPM) and runs it on my code. In general I would expect that they both give the same answer but that's not always the case. From time to time Pyright raises errors that Pylance does not, which breaks my automatic testing.

Is there a way to run Pyright in such a mode that it reasonably guarantees the same output as when using Pylance with VSCode?

**Comment by erictraut** [comment]
In general, pylance and pyright should give you the same answers. My team regularly uses both, and we haven't had any issues with discrepancies between the two.

There are two settings whose defaults differ between pylance and the pyright CLI:
1. `typeCheckingMode` defaults to `off` for pylance but defaults to `basic` for pyright. It sounds like you already have this enabled for pylance, so this shouldn't be an issue.
2. `useLibraryCodeForTypes` defaults to `true` for pylance but defaults to `false` for pyright. Inferring types from library code produces results that are good enough for completion suggestions, but you will generally not want to enable it for type checking. This is the most likely explanation for what you're seeing. If you haven't already done so, I recommend configuring pylance to turn off this setting.

The release cadence of pylance and pyright are not completely aligned. Pyright is generally published twice a week on a somewhat irregular schedule, and pylance is released once a week, typically on Wednesday. That means there's a (very small) possibility that a bug has been fixed or the behavior changed in pyright since the latest release of pylance. It's unlikely that this explains what you're seeing unless you're running an out-of-date version of pyright. If your github action is pinned to a specific version, make sure you're updating it regularly. If the github action is always using the latest version, you should be fine.

If those don't explain what you're seeing, please post more details here, and we can work to further diagnose it.

**Comment by albireox** [comment]
I tried the `useLibraryCodeTypes` parameter but that doesn't seem to make a difference. Right now the difference I'm seeing is that in an attribute that can be a class instance or `None`, when I try to use a method of that class I get `XXX is not a known member of "None"` in Pyright but everything is fine with Pylance with `typeCheckingMode: "basic"`. This is actually a case in which Pyright is probably right, but the code has been there for a long time and neither Pyright or Pylance have ever complained.

My GitHub Action doesn't pin Pyright, it just gets the latest release. Probably the issue is the different release times. In the past when I have had this issue sometimes it has solved by itself as Pylance or Pyright got update. In other cases I have had to fix the Pyright issue (which is not in itself bad, since usually it reports a real issue).

So, not a big issue, just wondering if there was a way to keep things a bit more consistent.

**Comment by erictraut** [comment]
Yes, this is a behavior that we changed in pyright after last week's pylance release. The release notes for pylance each week notes which version of pylance it used. If you want to keep things 100% consistent, you would need to (re)pin the version each week on Wed afternoon after a new version of pylance is released.

**Comment by albireox** [comment]
Sounds good. I think it's not worth pinning Pyright in all my Actions every week since 95% of the times this is not an issue, but it's good to know that this is expected in this case.

**Comment by erictraut** [comment]
Yes, this particular change was a bit of an anomaly. We made this change only after much internal debate and based on a bunch of [pyright user feedback](https://github.com/microsoft/pyright/discussions/1969). We decided that it was best to roll out the change in pyright prior to pylance, giving us an opportunity to get feedback from the smaller pyright user base prior to releasing to the ~2M pylance users. Apologies for any inconvenience this caused.
